### DISCUSSION TITLE
Contextive knowledge sharing outside of development environments

### DISCUSSION THREAD

**Discussion by FH-Inway** [dev-cycles/contextive#71] [Ideas]
In [Ubiquitous F# in Contextive, the Ubiquitous Language Tool | fsharpConf 2023](https://www.youtube.com/watch?v=bf1as7nTFEY&t=676s), @chrissimon-au shared a vision where Contextive is not just a tool for sharing knowledge in development environments. Instead, it acts as a core for a whole ecosystem of tools that enable the continued exchange of knowledge with a ubiquitious language.

Examples include a Slack chat bot, a browser extension and an Office 365 integration. A proof of concept demo is shown in the video for the Slack chat bot.

In the discussion of #28 , proof of concept implementations for all three examples were mentioned. This discussion is intended to continue this discussion.

Have fun! 

![image](https://github.com/user-attachments/assets/9b6ae517-4c80-4e5e-9a35-d4373000f9dd)

**Comment by chrissimon-au** [comment]
Thanks for setting this up @FH-Inway!

I've been thinking about this topic for a while, but there are a lot of options/tradeoffs to consider.

I've setup a survey to help capture structured data on folks preferences.  Feedback on the survey itself welcome in this discussion as well, but in the first instance, if anyone interested in this topic could fill out the survey, that'd be super helpful!

https://forms.gle/NgSahheaQg1FLX2w7

After some initial discussions, I'll start preparing a draft roadmap.

**Reply by chrissimon-au** [reply]
Survey result summary here: https://docs.google.com/forms/d/1BQsXmBZM7ORjPb3_H74elLgWcY6zgtGYQ1L8ftIlWqA/viewanalytics

**Comment by chrissimon-au** [comment]
Back onto this topic... only had a couple of survey respones, but I think enough to formulate a plan.

Here's a rough uncommitted roadmap for getting started on this work:

1. #54 - it represents a major change to the configuration approach, so would prefer to get started on expanding beyond the IDE after this is done, and it will simplify more flexible ways of identifying which context is relevant in a given usage scenario
2. #73 - see ticket for details on how it's proposed to configure it, including how to identify which context is relevant on a given page
3. #43 -  once the above are done, it should be straightforward to complete this, which would then offer the option for all the terms to be sourced and hosted externally

Beyond that, the 'later' pile to be prioritised once getting through the above, would include:
* #72
* Prod-ready chatbots (slack, teams, etc.)
* Mechanisms to support updating (or suggesting updates) to definitions from other locations

**Comment by chrissimon-au** [comment]
Hi everyone, it's been a while, but happy to say some great progress on this - since v1.16.0 released just now, you can now import glossaries from remote URLs and since v1.15.0 a few days ago you can split your terms across multiple glossary files and then they will get applied when working in files in the same folder or children.

There is now a lot more flexibility for organising these things, as well as a pathway for importing terminology into contextive from external sources.

See https://docs.contextive.tech/community/v/1.16.0/guides/setting-up-glossaries/#terms-defined-in-an-external-data-source for more details.

Will leave this discussion open until more progress has been made on the browser plugin, hopefully soon!

**Reply by FH-Inway** [reply]
Nice work :rocket:

Also like the new (well, new to me) documentation site at https://docs.contextive.tech/ . You might want to give it a more prominent place on the Github main page (e.g. the link in the top right, mention it earlier in the Readme) and fix the links to it in the Readme (currently getting a 404 for the "Getting started" links).

**Reply by chrissimon-au** [reply]
Thanks! Yes, the docs site is relative new (last few weeks) - thanks for picking up those 404s, a few links got lost in some early reshuffling as I was experimenting with the structure of the docs site.  Should be fixed now, including a more obvious link to docs at the top of the readme.

**Comment by chrissimon-au** [comment]
Folks, very happy to share that there's been some real progress on this! The Contextive Browser Extension is now available for Chromium based browsers.

See https://docs.contextive.tech/community/guides/installation/#web-browsers for documentation on installing and https://docs.contextive.tech/community/guides/setting-up-glossaries/#web-browsers for guidance on setting it up to make use of your existing glossary yml files.

A few things to note about this first release:

* It can only use glossary files  hosted on unauthenticated urls.  These could be private IPs on your private network, or even localhost if you have a local web server, but it doesn't support authentication.
* It does not yet support the `import` key, so you must point it at the yml files that contain the actual terminology
* Not exhaustively tested across multiple chromium browsers or websites.  If you find it causing an issue, you can deactivate each configured 'workspaces' (see the docs linked above for details)
* Not tested for performance with very large (> 10,000 terms) glossaries or very large web pages.
* Each user has to configure the extension individually (a future release will allow to export/import configuration or to load configuration from a URL)

The following screenshots show it in action on the contextive docs site, once configured with the sample 'Cargo' glossary file:

Highlighting the term `Cargo` with the blue/red underline:

<img width="965" height="158" alt="image" src="https://github.com/user-attachments/assets/2782ff89-ebd9-4b00-8304-c804cca443b4" />

Hovering over `Cargo`:

<img width="344" height="196" alt="image" src="https://github.com/user-attachments/assets/8d85388a-df7c-44b9-bf6d-300ea05517a6" />

Hovering over `Policy` (defined in multiple contexts):

<img width="398" height="304" alt="image" src="https://github.com/user-attachments/assets/4ccc41e4-b640-4570-9ac5-6287bf642afd" />

Viewing `Policy` in the glossary side panel - opened by clicking 'View in glossary' on the hover:

<img width="369" height="641" alt="image" src="https://github.com/user-attachments/assets/935e4e4a-028f-4525-8a26-73186cad3689" />

Would be very grateful for some testing and feedback if anyone gets a chance to try it out!  Thanks :)

**Reply by jdebarochez** [reply]
Hey Chris! 
I gave it a try and I had a hard time to set it up correctly (disclaimer, I'm really clumsy). I would love to have some visual feedback when:
- the glossary is not properly loaded
- the site is part of the workspace
I did a few back and forth to correctly link the glossary. First it was in `.txt`, then maybe badly indented, then I gave the wrong link. 😅
Then I could not know how to "reload" if ever I had to. Refreshing the page did the trick. I believe we should have a mention how to update the glossary and see the changes.

Here's my example workspace:
- my glossary is https://gist.github.com/jdebarochez/69730f3f9097300a352a3e2fdde28b70
- the URLs I allowed are:
  - https://www.zalando.fr/* ([shoes](https://www.zalando.fr/chaussures-homme/) page)
  - https://en.wikipedia.org/* ([shirt](https://en.wikipedia.org/wiki/Shirt) page)
  - https://gist.github.com/jdebarochez/69730f3f9097300a352a3e2fdde28b70 it's interesting to see it works on itself 😁 

<img width="1914" height="1015" alt="image" src="https://github.com/user-attachments/assets/43180105-eba3-4960-969d-6aad56f8956f" />

I find this 👆 could be really good use case when reading a pull-request on GitHub. 👍

**Reply by chrissimon-au** [reply]
@jdebarochez - thanks so much for giving it a try, and yes that is excellent feedback.  It was on my mind to give more visual indicator when there is an issue loading or parsing the glossary, but this is a good prompt to push that higher on the priority list :) #110 has been setup for this.

Also with forcing a retry/refresh of the glossary loading - #111 has been setup for this.

Great to hear that you see the utility of the use case for developers reviewing PRs!

And no need for the disclaimer, not clumsy at all! This is a very first version which is missing many of the UX affordances that would be nice, so thank you for persevering and getting it working! 🤩

**Comment by chrissimon-au** [comment]
@jdebarochez - v0.3 of the browser extension is now out and includes more visual indicators of the glossary loading status and the whether or not the extension is active on the current page.

<img width="364" height="250" alt="image" src="https://github.com/user-attachments/assets/eb851f47-d484-4ebb-bb8d-cdd63a61e0bb" />

<img width="364" height="358" alt="image" src="https://github.com/user-attachments/assets/7e5a1daf-9642-4930-8862-df978737a843" />

<img width="364" height="250" alt="image" src="https://github.com/user-attachments/assets/0173d673-c1b4-4bf4-be9b-9c7ea40ff146" />

Note that there is still an issue that's been identified if you configure contextive while on the page you want it to be active on - it doesn't activate until the page is reloaded or opened in a new tab.  This is being tracked now here: https://github.com/dev-cycles/contextive/issues/114
